src/FakeItEasy/Core/ArgumentEqualityComparer.cs

```diff
@@ -54,10 +62,33 @@ public bool AreEqual(object expectedValue, object argumentValue)
                     return string.Equals(expectedString, argumentValue);
                 }
 
-                return argumentValue is IEnumerable argumentEnumerable &&
-                    Enumerable.SequenceEqual(
-                        ((IEnumerable)expectedValue).Cast<object?>(),
-                        argumentEnumerable.Cast<object?>());
+                if (argumentValue is not IEnumerable argumentEnumerable)
+                {
+                    return false;
+                }
+
+                var expectedEnumerator = ((IEnumerable)expectedValue).GetEnumerator();
+                var actualEnumerator = argumentEnumerable.GetEnumerator();
+                var anyMoreExpected = expectedEnumerator.MoveNext();
+                var anyMoreActual = actualEnumerator.MoveNext();
+
+                while (anyMoreExpected && anyMoreActual)
+                {
+                    if (!this.areEqualElements(expectedEnumerator.Current, actualEnumerator.Current))
+                    {
+                        return false;
+                    }
+
+                    anyMoreExpected = expectedEnumerator.MoveNext();
+                    anyMoreActual = actualEnumerator.MoveNext();
+                }
+
+                if (anyMoreExpected ^ anyMoreActual)
+                {
+                    return false;
+                }

```

Comment:
Couldn't this be replaced by a call to the `SequenceEqual` overload that takes an equality comparer? of course we'd need an adapter from `IArgumentEqualityComparer` to `IEqualityComparer<T>`...

Reply:
Oh, yes. That's a much better idea.

Reply:
Doing now.